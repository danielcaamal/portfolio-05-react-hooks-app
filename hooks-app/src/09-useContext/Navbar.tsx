import { Link, NavLink } from "react-router-dom";

export const Navbar = () => {

    return (
        <nav className="navbar navbar-expand-lg navbar-dark bg-dark rounded-3">
            <div className="container-fluid">
                <Link className="navbar-brand" to="/">UseContext</Link>
                <div className="collapse navbar-collapse" id="navbarNavAltMarkup">
                    <div className="navbar-nav">
                        <NavItem to="/">Home</NavItem>
                        <NavItem to="/login">Login</NavItem>
                        <NavItem to="/about">About</NavItem>
                    </div>
                </div>
            </div>
        </nav>
    );
}

const NavItem = ({ children, to } : {children: any, to:string}) => {
    return (
        <NavLink 
            className={ (args) => `nav-item nav-link ${args.isActive && 'active' }` }
            to={to}>
            {children}
        </NavLink>
    );
}