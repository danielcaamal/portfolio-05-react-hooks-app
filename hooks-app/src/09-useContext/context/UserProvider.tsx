import { useState } from "react";
import { UserContext } from "./UserContext";


export const UserProvider = ({ children } : { children: JSX.Element[] }) => {

    const [user, setUser] = useState({});

    const setUserContext = (user: any) => {
        setUser(user);
    }

    return (
        <UserContext.Provider value={{ user, setUserContext }}>
            { children }
        </UserContext.Provider>
    );
}