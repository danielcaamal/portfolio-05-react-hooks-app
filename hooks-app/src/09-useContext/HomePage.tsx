import { useContext } from "react";
import { UserContext } from "./context/UserContext";


export const HomePage = () => {
    const { user, setUserContext } = useContext(UserContext);

    return (
        <>
            <h1
                id='home-page-title'
                >HomePage: {user?.name}</h1>
            <hr />
            <pre id='home-page-pre'>
                { JSON.stringify(user, null, 3) }
            </pre>
            <button
                id='home-page-button'
                type="button" 
                className="btn btn-primary"
                onClick={ () => setUserContext({ id: 123, name: 'John Doe', email: 'email@email.com' }) }
                >Set User
            </button>
        </>
    );
};