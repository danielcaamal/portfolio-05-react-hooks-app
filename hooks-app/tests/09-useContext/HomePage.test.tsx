import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { HomePage } from '../../src/09-useContext/HomePage';
import { UserContext } from '../../src/09-useContext/context/UserContext';

describe('HomePage', () => {
    it('should render', () => {
        const { container } = render(
            <UserContext.Provider value={{ user: null }}>
                <HomePage />
            </UserContext.Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with user', () => {
        const value = { user: { name: 'John' } };
        const { container } = render(
            <UserContext.Provider value={value}>
                <HomePage />
            </UserContext.Provider>
        );
        expect(container).toMatchSnapshot();
        const pre = container.querySelector('#home-page-pre') as HTMLPreElement;
        expect(pre.textContent).toContain(value.user.name)
    });

    it ('should set user', () => {
        const value = { user: null, setUserContext: jest.fn() };
        const { container } = render(
            <UserContext.Provider value={value}>
                <HomePage />
            </UserContext.Provider>
        );
        const button = container.querySelector('#home-page-button') as HTMLButtonElement;
        fireEvent.click(button);
        expect(value.setUserContext).toHaveBeenCalled();
    });
});