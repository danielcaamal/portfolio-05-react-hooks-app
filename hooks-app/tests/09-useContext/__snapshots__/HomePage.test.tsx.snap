// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should render 1`] = `
<div>
  <h1
    id="home-page-title"
  >
    HomePage: 
  </h1>
  <hr />
  <pre
    id="home-page-pre"
  >
    null
  </pre>
  <button
    class="btn btn-primary"
    id="home-page-button"
    type="button"
  >
    Set User
  </button>
</div>
`;

exports[`HomePage should render with user 1`] = `
<div>
  <h1
    id="home-page-title"
  >
    HomePage: 
    John
  </h1>
  <hr />
  <pre
    id="home-page-pre"
  >
    {
   "name": "John"
}
  </pre>
  <button
    class="btn btn-primary"
    id="home-page-button"
    type="button"
  >
    Set User
  </button>
</div>
`;
